doc/design/upgrade.md

```diff
@@ -0,0 +1,118 @@
+
+vic-machine upgrade cannot upgrade VCH to a version newer than itself.
+
+### Impact
+
+The VCH will have downtime during upgrade, suppose container lifecyle is not closely coupled with VCH up time, for VIC has good network isolation. User could run appliacation and connect to their container without VCH is running. But there still have some impact, including docker attach, docker log and docker port mapping, which can be resumed after upgrade is finished.
+
+In the future, will consider to add live migration, so user does not experience service broken during upgrade. Integration docker log with syslogd is another option to reduce impact of VCH downtime.
+
+### VCH status
+
+vic-machine ls and inspect should be able to show the VCH versions and upgrade status.
+
+### Roll Back
+
+For anything wrong happens during upgrade, vic-machine will rollback any changes made to have VCH return to original version and status. If VC/vSphere connection is broken during upgrade, the upgrade or rollback will fail, but user should be able to rollback previously broken upgrade, to recover VCH.
+
+### Downgrade
+
+For any running VCH, VIC admin can downgrade existing VCH to an old version. (I assume this function is only used for upgrade roll back, we might not want to recommend downgrade for anything working well, then the downgrade scope will be same to upgrade versions)
+
+Verified docker-engine upgrade/downgrade, which works well between 1.11.2 and 1.12. The running container will be stopped after upgrade/downgrade, but is good to start again in new version. And all images and containers information is not lost. So from technical point of view, we should support same story.
+
+### Internet connection
+Internet connection is not required to upgrade/downgrade VCH, but newer version's binary should be available for vic-machine.
+
+## Implementation/Issues
+
+### Versioning
+
+#### Freeze following attributes definition in VCH configuration and container VM configuration
+- VCH configuration: ID, Name, Version
+- Container VM serial port configuration.
+- Container VM log/debug file location.
+
+#### Configuration Version
+
+Both VCH and container configuration will have version and the value is same with what vic-machine version command shown. After upgrade, VCH configuration version should be updated to new version, but container VM configuration version will still be old one.
+
+New version's VCH will work with both old and new versions containers.
+
+#### Embed iso file version
+
+vic-machine need to identify the iso file version before and after upgrade. Two options here:
+- Add version into iso file Primary Volume Descriptor, and read back from vic-machine
+- Leverage iso file name including version, e.g. appliance-0.5.0.iso
```

Comment:
I didn't mean for the ISO names (although I'm struggling to think of a reason _not_ to do it there), but for the guestinfo configuration - that's what the UI plugin looks at, and how vic-machine determines that a VM is a VCH.
If the same config mechanism is used for other products then we'll need a way to filter.